### PR TITLE
[CodeCamp #1503] Add InstanceSegMetric to MMEval

### PR DESCRIPTION
**Pull request by Pzzzzz5142**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add InstanceSegMetric to MMEval

## Modification

+ mmeval/metrics/instance_seg.py
+ mmeval/metrics/utils/evaluate_semantic_instance.py
+ mmeval/metrics/utils/util_3d.py
+ tests/test_metrics/test_instance_seg.py

### CODE REVIEW

File: mmeval/metrics/utils/evaluate_semantic_instance.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# adapted from https://github.com/ScanNet/ScanNet/blob/master/BenchmarkScripts/3d_evaluation/evaluate_semantic_instance.py # noqa
+import numpy as np
+from copy import deepcopy
+
+from . import util_3d
```
Comment (ice-tong) [inline mmeval/metrics/utils/evaluate_semantic_instance.py:6]: This script seems adapted from `https://github.com/ScanNet/ScanNet`.
Copy it as it is and put it in `mmeval/metrics/_vendor`

File: mmeval/metrics/utils/util_3d.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# adapted from https://github.com/ScanNet/ScanNet/blob/master/BenchmarkScripts/util_3d.py # noqa
```
Comment (ice-tong) [inline mmeval/metrics/utils/util_3d.py:2]: Copy it as it is and put it in `mmeval/metrics/_vendor`

File: mmeval/metrics/utils/evaluate_semantic_instance.py (hunk 1 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/utils/evaluate_semantic_instance.py]: moved

File: mmeval/metrics/utils/util_3d.py (hunk 2 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/utils/util_3d.py]: moved

File: mmeval/metrics/instance_seg.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+
+def instance_seg_eval(gt_semantic_masks,
+                      gt_instance_masks,
+                      pred_instance_masks,
+                      pred_instance_labels,
+                      pred_instance_scores,
+                      valid_class_ids,
+                      class_labels,
+                      options=None):
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:73]: It is recommended to write this function as a class method～

File: mmeval/metrics/instance_seg.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            file_name = f'{id}_{i}'
+            info[file_name] = dict()
+            info[file_name]['mask'] = (mask == i).astype(int)
+            info[file_name]['label_id'] = valid_class_ids[label[i]]
+            info[file_name]['conf'] = score[i]
+        infos.append(info)
+    return infos
+
+
+def rename_gt(gt_semantic_masks, gt_instance_masks, valid_class_ids):
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:37]: It is recommended to write this function as a class method～

File: mmeval/metrics/instance_seg.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+    Args:
+        masks (list[numpy.ndarray]): Per scene predicted instance masks.
+        labels (list[numpy.ndarray]): Per scene predicted instance labels.
+        scores (list[numpy.ndarray]): Per scene predicted instance scores.
+        valid_class_ids (tuple[int]): Ids of valid categories.
+
+    Returns:
+        list[dict]: Per scene aggregated predictions.
+    """
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:21]: It is recommended to write this function as a class method～

File: mmeval/metrics/instance_seg.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            ...
+        }
+    }
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        assert self.dataset_meta is not None
+        self.classes = self.dataset_meta['classes']
+        self.valid_class_ids = self.dataset_meta['seg_valid_class_ids']
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:210]: We should support accepting the classes as init parameter～

File: mmeval/metrics/instance_seg.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+            self._results.append((deepcopy(prediction), deepcopy(groundtruth)))
+
+    def compute_metric(self, results: List[List[Dict]]) -> Dict[str, float]:
+        """Compute the metrics from processed results.
+
+        Args:
+            results (list): The processed results of each batch.
+        Returns:
+            Dict[str, float]: The computed metrics. The keys are the names of
+            the metrics, and the values are corresponding results.
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:241]:
```suggestion
        Args:
            results (list): The processed results of each batch.
 
        Returns:
            Dict[str, float]: The computed metrics. The keys are the names of
            the metrics, and the values are corresponding results.
```

File: tests/test_metrics/test_instance_seg.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import numpy as np
+import torch
+import unittest
+
+from mmeval.metrics import InstanceSeg
+
+
+class TestInstanceSegMetric(unittest.TestCase):
+
```
Comment (ice-tong) [inline tests/test_metrics/test_instance_seg.py:10]: Suggesting using pytest to unify the test style~

File: mmeval/metrics/instance_seg.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+            - pts_instance_mask(numpy.ndarray): Predicted instance masks.
+            - instance_labels(numpy.ndarray): Predicted instance labels.
+            - instance_scores(numpy.ndarray): Predicted instance scores.
+            groundtruths (Sequence[Dict]): A sequence of dict. Each dict
+                represents a groundtruths for an image, with the following
+                keys:
+            - pts_instance_mask(numpy.ndarray): Ground truth instance masks.
+            - pts_semantic_mask(numpy.ndarray): Ground truth semantic masks.
+        """
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:230]:
```suggestion
        which will be used to compute the metrics when all batches
        have been processed.
        
        Args:
            predictions (Sequence[Dict]): A sequence of dict. Each dict
                representing a detection result, with the following keys:

                - pts_instance_mask(numpy.ndarray): Predicted instance masks.
                - instance_labels(numpy.ndarray): Predicted instance labels.
                - instance_scores(numpy.ndarray): Predicted instance scores.
            
            groundtruths (Sequence[Dict]): A sequence of dict. Each dict
                represents a groundtruths for an image, with the following
                keys:
            
                - pts_instance_mask(numpy.ndarray): Ground truth instance masks.
                - pts_semantic_mask(numpy.ndarray): Ground truth semantic masks.
        """
```

File: mmeval/metrics/instance_seg.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+        for prediction, groundtruth in zip(predictions, groundtruths):
+            self._results.append((deepcopy(prediction), deepcopy(groundtruth)))
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:232]: Why we need deepcopy~

Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py:235]: If we don't use deepcopy, it will change the value of it which may cause confusion.

File: mmeval/metrics/instance_seg.py (hunk 3 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py]: fixed

File: mmeval/metrics/instance_seg.py (hunk 4 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py]: fixed

File: mmeval/metrics/instance_seg.py (hunk 5 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py]: fixed

File: tests/test_metrics/test_instance_seg.py (hunk 8 above)
Comment (Pzzzzz5142) [reply tests/test_metrics/test_instance_seg.py]: fixed

File: mmeval/metrics/instance_seg.py (hunk 6 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py]: fixed

File: mmeval/metrics/instance_seg.py [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+
+    def instance_seg_eval(self,
+                          gt_semantic_masks,
+                          gt_instance_masks,
+                          pred_instance_masks,
+                          pred_instance_labels,
+                          pred_instance_scores,
+                          valid_class_ids,
+                          class_labels,
+                          options=None):
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:246]: It is better to merge the implementation in this function into `compute_metric`, which avoid passing a lot of parameters~

File: mmeval/metrics/instance_seg.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+        pred_instance_masks = []
+        pred_instance_labels = []
+        pred_instance_scores = []
+
+        for result_pred, result_gt in results:
+            gt_semantic_masks.append(result_gt['pts_semantic_mask'])
+            gt_instance_masks.append(result_gt['pts_instance_mask'])
+            pred_instance_masks.append(result_pred['pts_instance_mask'])
+            pred_instance_labels.append(result_pred['instance_labels'])
+            pred_instance_scores.append(result_pred['instance_scores'])
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:172]: A suggestion that simplify the code~
```suggestion
        gt_semantic_masks = [gt['pts_semantic_mask'] for _, gt in results]
        gt_instance_masks = [gt['pts_instance_mask'] for _, gt in results]
        pred_instance_masks = [pred['pts_instance_mask'] for pred, _ in results]
        pred_instance_labels = [pred['instance_labels'] for pred, _ in results]
        pred_instance_scores = [pred['instance_scores'] for pred, _ in results]
```

File: mmeval/metrics/instance_seg.py [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+                'ap25%': 1.0
+            },
+            'chair': {
+                'ap': 1.0,
+                'ap50%': 1.0,
+                'ap25%': 1.0
+            },
+            ...
+        }
+    }
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:102]: invalid indent~

File: mmeval/metrics/instance_seg.py [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+        >>> from mmeval import InstanceSegMetric
+        >>> seg_valid_class_ids = (3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 14, 16, 24,
+        >>>                 28, 33, 34, 36, 39)
+        >>> class_labels = ('cabinet', 'bed', 'chair', 'sofa', 'table', 'door',
+        ...         'window', 'bookshelf', 'picture', 'counter', 'desk',
+        ...         'curtain', 'refrigerator', 'showercurtrain', 'toilet',
+        ...         'sink', 'bathtub', 'garbagebin')
+        >>> dataset_meta = dict(
+        ...     seg_valid_class_ids=seg_valid_class_ids, classes=class_labels)
+        >>>
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:32]: The example is too complicated and needs to be simplified~

File: mmeval/metrics/instance_seg.py [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+
+class InstanceSeg(BaseMetric):
+    """3D instance segmentation evaluation metric.
+
+    Args:
+        dataset_meta (dict, optional): Provide dataset meta information.
+        classes (List[str], optional): Provide dataset classes information as
+            an alternative to dataset_meta.
+        valid_class_ids (List[int], optional): Provide dataset valid class ids
+            information as an alternative to dataset_meta.
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:19]:
```suggestion
        valid_class_ids (List[int], optional): Provide dataset valid class ids
            information as an alternative to dataset_meta.
        **kwargs: Keyword parameters passed to :class:`BaseMetric`.
```

File: mmeval/metrics/instance_seg.py [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+from copy import deepcopy
+from typing import Dict, List, Optional, Sequence
+
+from mmeval.core.base_metric import BaseMetric
+from mmeval.metrics._vendor.scannet import scannet_eval
+
+
+class InstanceSeg(BaseMetric):
+    """3D instance segmentation evaluation metric.
+
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:13]: It is better to add a summary of this Metric. 
i.e. Introduction of ScanNet

File: mmeval/metrics/instance_seg.py [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+                warnings.warn('`classes` and `valid_class_ids` are ignored.')
+            self.classes = self.dataset_meta['classes']
+            self.valid_class_ids = self.dataset_meta['seg_valid_class_ids']
+        else:
+            if not isinstance(classes, list):
+                raise TypeError('`classes` should be List[str]')
+            if not isinstance(valid_class_ids, list):
+                raise TypeError('`valid_class_ids` should be List[int]')
+            self.classes = classes
+            self.valid_class_ids = valid_class_ids
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:124]: We can set a classes property to address this, please ref: https://github.com/open-mmlab/mmeval/blob/e635ba6338105d92f053b46320f3b1ca340ef9b4/mmeval/metrics/mean_iou.py#L114-L138

File: mmeval/metrics/instance_seg.py [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+                - pts_instance_mask(np.ndarray): Predicted instance masks.
+                - instance_labels(np.ndarray): Predicted instance labels.
+                - instance_scores(np.ndarray): Predicted instance scores.
+
+            groundtruths (Sequence[Dict]): A sequence of dict. Each dict
+                represents a groundtruths for an image, with the following
+                keys:
+
+                - pts_instance_mask(np.ndarray): Ground truth instance masks.
+                - pts_semantic_mask(np.ndarray): Ground truth semantic masks.
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:146]:
```suggestion
            predictions (Sequence[Dict]): A sequence of dict. Each dict
                representing a detection result, with the following keys:

                - pts_instance_mask (np.ndarray): Predicted instance masks.
                - instance_labels (np.ndarray): Predicted instance labels.
                - instance_scores (np.ndarray): Predicted instance scores.

            groundtruths (Sequence[Dict]): A sequence of dict. Each dict
                represents a groundtruths for an image, with the following
                keys:

                - pts_instance_mask (np.ndarray): Ground truth instance masks.
                - pts_semantic_mask (np.ndarray): Ground truth semantic masks.
```

File: mmeval/metrics/instance_seg.py [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+            unique = np.unique(instance_mask)
+            assert len(unique) < 1000
+            for i in unique:
+                semantic_instance = semantic_mask[instance_mask == i]
+                semantic_unique = np.unique(semantic_instance)
+                assert len(semantic_unique) == 1
+                if semantic_unique[0] < len(valid_class_ids):
+                    instance_mask[
+                        instance_mask ==
+                        i] = 1000 * valid_class_ids[semantic_unique[0]] + i
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:234]:
```suggestion
                    instance_mask[instance_mask==i] = 1000 * valid_class_ids[semantic_unique[0]] + i  #noqa: E501
```

File: mmeval/metrics/instance_seg.py [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+            valid_class_ids (tuple[int]): Ids of valid categories.
+
+        Returns:
+            list[np.array]: Per scene instance masks.
+        """
+        renamed_instance_masks = []
+        for semantic_mask, instance_mask in zip(gt_semantic_masks,
+                                                gt_instance_masks):
+            unique = np.unique(instance_mask)
+            assert len(unique) < 1000
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:226]: It is better to add a assert message:
```
assert len(unique) < 1000, "the xxx should large than 1000."
```

File: mmeval/metrics/instance_seg.py [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+            for i in range(n_instances):
+                file_name = f'{id}_{i}'
+                info[file_name] = dict()
+                info[file_name]['mask'] = (mask == i).astype(int)
+                info[file_name]['label_id'] = valid_class_ids[label[i]]
+                info[file_name]['conf'] = score[i]
+            infos.append(info)
+        return infos
+
+    def rename_gt(self, gt_semantic_masks, gt_instance_masks, valid_class_ids):
```
Comment (ice-tong) [inline mmeval/metrics/instance_seg.py:210]: The `rename` in function name `rename_gt` seems a bit weird, what is be renamed?

File: mmeval/metrics/instance_seg.py (hunk 12 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py:173]: In this way, we would do 5 for loops, which may be time consuming.

File: mmeval/metrics/instance_seg.py (hunk 21 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py:211]: It seems that this function rename the input ground truth in order to evaluate in ScanNet protocol provided by [ScanNet repo](https://github.com/ScanNet/ScanNet).

File: mmeval/metrics/instance_seg.py (hunk 11 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py]: fixed

File: mmeval/metrics/instance_seg.py (hunk 13 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py]: fixed

File: mmeval/metrics/instance_seg.py (hunk 16 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py:12]: fixed

File: mmeval/metrics/instance_seg.py (hunk 17 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py]: fixed

File: mmeval/metrics/instance_seg.py (hunk 14 above)
Comment (Pzzzzz5142) [reply mmeval/metrics/instance_seg.py]: It is hard to simplify it for this metric need many code to generate input.
